rare/components/tabs/games/game_widgets/__init__.py

```diff
@@ -0,0 +1,216 @@
+            # lk: base it on icon widgets, the two lists should be identical
+            icon_widgets = self._icon_container.findChildren(IconGameWidget)
+            list_widgets = self._list_container.findChildren(ListGameWidget)
+            icon_app_names = set([iw.game.app_name for iw in icon_widgets])
+            list_app_names = set([lw.game.app_name for lw in list_widgets])
+            games = self.core.get_game_list(update_assets=True)
+            game_app_names = set([g.app_name for g in games])
+            new_icon_app_names = game_app_names.difference(icon_app_names)
+            new_list_app_names = game_app_names.difference(list_app_names)
+            for app_name in new_icon_app_names:
+                game = self.core.get_game(app_name, update_meta=False)
+                iw = IconGameWidget(game, self._game_utils)
+                self._icon_container.layout().addWidget(iw)
+            for app_name in new_list_app_names:
+                game = self.core.get_game(app_name, update_meta=False)
+                lw = ListGameWidget(game, self._game_utils)
+                self._list_container.layout().addWidget(lw)
+            self.sort_list()
+
+    @pyqtSlot(list)
+    def on_install(self, app_names: List[str]):
+        for app_name in app_names:
+            iw, lw = self.__find_widget(app_name)
+            if iw is not None:
+                iw.is_installed = True
+            if lw is not None:
+                lw.is_installed = True
+        self.sort_list()
+
+    @pyqtSlot(str)
+    def on_uninstall(self, app_name: str):
+        iw, lw = self.__find_widget(app_name)
+        if iw is not None:
+            iw.is_installed = False
+        if lw is not None:
+            lw.is_installed = False
+        self.sort_list()
+
+    @pyqtSlot(str)
+    def on_verified(self, app_name: str):
+        iw, lw = self.__find_widget(app_name)
+        if iw is not None:
+            iw.needs_verification = False
+        if lw is not None:
+            lw.needs_verification = False
+
+    @pyqtSlot(str)
+    def on_image_completed(self, app_name: str):
+        iw, lw = self.__find_widget(app_name)
+        if iw is not None:
+            iw.setAppNamePixmap(app_name, iw.is_installed)
+        if lw is not None:
+            lw.setAppNamePixmap(app_name, lw.is_installed)
+
+    # lk: this should go in downloads and happen once
+    def __find_game_for_dlc(self, app_name: str):
+        game = self.core.get_game(app_name, False)
+        # lk: how can an app_name not refer to a game?
+        if not game:
+            return None
+        if game.is_dlc:
+            game_list = self.core.get_game_list(update_assets=False)
+            game = list(
+                filter(
+                    lambda x: x.asset_infos["Windows"].catalog_item_id
+                    == game.metadata["mainGameItem"]["id"],
+                    game_list,
+                )
+            )
+            return game[0].app_name
+        return app_name
```

Comment:
Missing typing! Yet another time, you probably have a rule for that which I missed.